staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -517,7 +526,7 @@ func (r Request) finalURLTemplate() url.URL {
 			segments[index+3] = "{name}"
 		}
 	}
-	url.Path = path.Join(segments...)
+	url.Path = strings.Join(segments, "/")
```

Comment:
why do we change this line?

Reply:
@caesarxuchao, @roycaihw pointed out that in this test [0], the expected full url for the uri `base/url` should contain the `/` in the `base/url`. Let me know if this is not the right assumption.

[0]
```
Request: NewRequest(nil, "DELETE", uriBaseURL, "", ContentConfig{GroupVersion: &schema.GroupVersion{Group: "test"}}, Serializers{}, nil, nil, 0).
				Prefix("/api/v1/namespaces/ns/r3/name1"),
ExpectedFullURL:  "/base/url/api/v1/namespaces/ns/r3/name1",
ExpectedFinalURL: "/base/url/api/v1/namespaces/%7Bnamespace%7D/r3/%7Bname%7D",
```

Reply:
Got it.

But I'm confused why we are testing the uriSingleSlash and uriBaseURL in the first place. They are invalid URL, URL needs a non-empty scheme.

Reply:
to be clear, the usage of `finalURLTemplate` is to only replace actual namespace and resource name into `{namespace}` and `{name}` to solve the cardinality problem. no other part of the url should be modified.

I value the testcases more than the implementation, because the current implementation is heavily based on string parsing and assumptions about kubernetes API convention, which is brittle to future changes and corner cases

a better fix could be forcing how rest client is used (e.g. by dynamic client). Ref: https://github.com/kubernetes/kubernetes/issues/68690

> They are invalid URL, URL needs a non-empty scheme

@caesarxuchao good point. Perhaps we should add a validation in rest client when a request is composed

Reply:
I agree that the testcases are good to add and that this implementation seems convoluted. Like why not check the prefix against the possible prefixes rather than the complicated logic of figuring out which section has the group in it?

e.g. `if strings.HasPrefix(path, "/api/") { ... } else if string.HasPrefix(path, "/apis/") { ... }`

I'd also like to stick with path-specific functions if possible, as Clayton noted.

Reply:
@lavalamp are you suggesting that this function should be really cleaned-up in this PR? I do think it's very hard to read and there is commented out code also. 🐉

Reply:
Uh, I don't know if I'd block on making it perfect (and please don't mix refactoring and feature-adding in the same commit!), but at least the new portions of it shouldn't make it worse (e.g. don't stop using path.Join).